src/auth/swt-strategy.ts

```diff
@@ -0,0 +1,84 @@
+import { Request, Response, NextFunction } from "express";
+import { SiweError, SiweErrorType } from 'siwe';
+import { SignatureTokenDecoder } from "./signature-token-decoder";
+import { SignatureValidator } from "./signature-validator";
+
+
+// TODO: siwtch to authorization header
+const SIWE_AUTH_HEADER = 'x-signature';
+
+export interface SwtStrategyOptions {
+    skipOnMissingSignature?: boolean;
+    allowedOrigins: Array<string | RegExp>; 
+}
+
+/** SWT Strategy: SIWE authentiacation strategy that uses JWT-like approach for authneticaiton */
+export class SwtStrategy {
+
+    constructor(
+        private validator: SignatureValidator,
+        private decoder: SignatureTokenDecoder,
+        private options: SwtStrategyOptions
+    ) {}
+
+    async authenticate(req: Request, res: Response, next: NextFunction) {
+        const signatureToken = req.header(SIWE_AUTH_HEADER);
+
+        if(!signatureToken) {
+            if(!this.options.skipOnMissingSignature) {
+                throw new Error('Signature not provided');   
+            }
+            console.log('Signature not provided, will skip');
+            return next();
+        }
+        
+        try {
+            console.log('Signature found, will decode and validate');
+            const {payload, signature} = this.decoder.decode(signatureToken);
+
+            const uri = getOriginalDomain(req);
+            req.wallet = await this.validator.validateSignature(payload, signature, uri, this.options.allowedOrigins, new Date());
+            console.log('Signature is valid', req.wallet);
+            next();
```

Comment:
Not sure about having next() in try block.
If next middleware will throw an exception, it will be logged as "Error during authentication" which is not correct.
Might be move it after catch and handle such error separatelly?

Reply:
looks like next() cannot throw, so, nevermind